### PR TITLE
make console/tty write safe

### PR DESCRIPTION
**Pull request by zhaozg**
make p and print output full/safe.

### CODE REVIEW

File: deps/pretty-print.lua [hunk 1]
```diff
@@ -297,6 +297,16 @@ function dump(value, recurse, nocolor)
...
 end
 
+local function console_write(fs, s)
+  s = s .. '\n'
+  repeat
+    n = uv.try_write(fs, s)
+    if n then
+      s = s:sub(n+1)
+    end
+    n = 0
```
Comment (rphillips) [inline deps/pretty-print.lua]: next line does a comparison on n equaling the length of s, but n holds the write length. There needs to be a counter.

**Comment by squeek502** [comment]
I think using `try_write` will make it possible for output to be truncated/dropped which is a worse problem than infinite loops causing memory exhaustion. Is there a different problem this is trying to solve? If so, could you provide a reproduction of that problem?

See the comments here: https://github.com/luvit/luvit/issues/1092#issuecomment-575507954

File: deps/pretty-print.lua (hunk 1 above)
Comment (zhaozg) [reply deps/pretty-print.lua]: s = s:sub(n+1) will skip data has been write, all write will get s=''

**Comment by zhaozg** [comment]
>I think using try_write will make it possible for output to be truncated/dropped which is a worse problem than infinite loops causing memory exhaustion. Is there a different problem this is trying to solve? If so, could you provide a reproduction of that problem?

I found uv_write will runcated/dropped data, when process:exit(), but try_write make all output. I almost forget about https://github.com/luvit/luvit/issues/1092#issuecomment-575507954.

**Comment by squeek502** [comment]
Using `try_write` for `tty` makes some sense to me, but I feel like I don't have enough knowledge to know what the trade-offs are here.

@SinisterRectus? @rphillips? Thoughts?

**Comment by SinisterRectus** [comment]
I looked at this in Node.js and I *think* they use try_write for tty.

**Comment by SinisterRectus** [comment]
Looking at why we reroute `print`:
>Print replacement that goes through libuv.  This is useful on windows
>to use libuv's code to translate ansi escape codes to windows API calls.

Something I just discovered (in Luvi 2.16.0 on Windows) is that if you open stdio handles with luv, Lua's IO calls are affected. Interestingly, whether you open fd 2 or 3 or both, Lua's io.stdout and io.stderr are both affected. I don't understand enough about how this works under the hood to know what's going on here. Can we rely on this behavior instead of re-writing `print`?

```lua
local uv = require('uv')

local colors = {
  black   = 30,
  red     = 31,
  green   = 32,
  yellow  = 33,
  blue    = 34,
  magenta = 35,
  cyan    = 36,
  white   = 37,
}

local function colorize(str, color)
  return string.format('\27[%i;%im%s\27[0m', 1, color, str)
end

local function open(fd)
  local handle
  if uv.guess_handle(fd) == 'tty' then
    handle = assert(uv.new_tty(fd, true))
  else
    handle = uv.new_pipe(false)
    uv.pipe_open(handle, fd)
  end
  return handle
end

open(0)
open(1)
open(2)

print(colorize('print', colors.cyan))
io.stdout:write(colorize('io.stdout:write', colors.green), '\n')
io.stderr:write(colorize('io.stderr:write', colors.yellow), '\n')
error(colorize('error', colors.red))
```
image

**Comment by squeek502** [comment]
@SinisterRectus that doesn't seem to be universal. Here's what I get on Windows 7:

wincolors

Do you get the colors if you just use print? I think at some point the Windows console added native support for certain ANSI escape sequences.

**Comment by SinisterRectus** [comment]
No. Without initializing the tty handles, I get the same raw output on Windows 10 that you see there in Windows 7. Good to know that it is not universal.

**Comment by zhaozg** [comment]
May I merge this? @squeek502 @rphillips @SinisterRectus

File: deps/pretty-print.lua [hunk 2]
```diff
@@ -297,6 +297,26 @@ function dump(value, recurse, nocolor)
...
+    repeat
+      local n, e = uv.try_write(fs, s)
+      if n then
+        s = s:sub(n+1)
+        n = 0
+      else
+        if e:match('^EAGAIN') then
+          n = 0
+        else
+          assert(n, reason)
```
Comment (squeek502) [inline deps/pretty-print.lua]: Where is `reason` coming from? Should this be `e`?

**Comment by squeek502** [comment]
I'm okay with trying this.

File: deps/pretty-print.lua (hunk 2 above)
Comment (zhaozg) [reply deps/pretty-print.lua]: Yes

**Pull request merged**